File: python/002. Add Two Numbers.py
#coding=utf-8
__author__ = 'xuxuan'
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        return Solution().addTwo(l1,l2,False)
    def addTwo(self,l1,l2,flag):
        t=False
        val=0;
        if(l1):
            val+=l1.val
            t=True
        if(l2):
            val+=l2.val
            t=True
        if(flag):
            val+=1
            t=True
        if(val>=10):
            flag=True
            val-=10
        else :
            flag=False
        if(t):
            ans=ListNode(val)
            ans.next=Solution().addTwo(l1.next if l1 else None ,l2.next if l2 else None,flag)
            return ans
        else :
            return None


a=ListNode(0)
b=ListNode(0)
sol=Solution()
print sol.addTwoNumbers(a,b)

File: js/002. Add Two Numbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    return add(l1,l2,false)
};
var add =function(l1,l2,flag) {
	if (l1===null) {
		if(flag){
			return add(new ListNode(1),l2,false)
		}
		return l2
	}
    if(l2===null){
    	if(flag){
			return add(l1,new ListNode(1),false)
		}
		return l1
    }
    var v=l1.val+l2.val+(flag?1:0);
    var ans=new ListNode(v%10);
    ans.next=add(l1.next,l2.next,v>=10);
    return ans;
};